### ISSUE TITLE
Static type safety for the union types

### ISSUE THREAD

**Issue by zhyatt** [nanocurrency/nano-node#2117]
C++ doesn't have opaque typedefs, which means we can mix account / block_hash "types" (which are just aliases to uint256), etc, which gives room for mistakes and confusing code.

This idiom, an anonymous union, was suggested which should have zero overhead:

```class uint512
{
public:
union
{
int x;
int y;
};
};


class qualified_root : public uint512
{
};


void func (qualified_root root);


int main ()
{
qualified_root root;
root.x;
root.y;


uint512 raw;
// func (raw); // not allowed
func (root);
}
```

Mention of the boost strong typedef was also made, and looks quite useful, but in this case we could also move some of the methods like "to_account ()" to the nano::account sub-class to keep related methods contained, which could be beneficial. Open to other options for this.

**Comment by cryptocode** [comment]
Relevant (boost strong typedef author chimes in): https://groups.google.com/a/isocpp.org/forum/#!topic/std-proposals/GP2Crh9hY84

**Comment by zhyatt** [comment]
Resolved by #2300
